RecoLocalCalo/HGCalRecAlgos/src/HGCalESProducerTools.cc

```diff
@@ -0,0 +1,95 @@
+#include "RecoLocalCalo/HGCalRecAlgos/interface/HGCalESProducerTools.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/RegexMatch.h"
+#include <sstream>  // for std::istringstream
+#include <regex>
+
+namespace hgcal {
+
+  // @short search first match to a given typecode key in a JSON (following insertion order)
+  // allow glob patterns, e.g. 'ML-*', 'M[LH]-[A-Z]3*', etc.
+  std::string search_modkey(const std::string& module, const json& data, const std::string& name = "") {
+    if (!data.is_object() or data.empty()) {
+      cms::Exception ex("InvalidData");
+      ex << "'" << name << "' does not have the expected map/dict structure!";
+      ex.addContext("Calling hgcal::search_modkey()");
+      throw ex;
+    }
+    for (auto it = data.begin(); it != data.end(); ++it) {
+      std::regex re(edm::glob2reg(it.key()));
+      if (std::regex_match(module, re)) {  // found matching key !
+        edm::LogInfo("search_modkey") << "search_modkey: Matched module='" << module << "' to modkey='" << it.key()
+                                      << "'";
+        return it.key();  // return matching key
+      }
+    }
+    edm::LogError("search_modkey") << "search_modkey: Could not find matching key for '" << module << "' in '" << name
+                                   << "'! Returning first key '" << data.begin().key() << "'...";
+    return data.begin().key();  // no matching key found in whole JSON map
```

Comment:
Why does it return the first key, in this case, instead of throwing an exception?

Reply:
Unfortunately most `LogError` messages are basically ignored in production workflows, if not outright suppressed.

Is it meaningful for the client code to continue using the first key instead of a matching key ?
Or is it an error condition that should be checked and handled explicitly ?

Reply:
Okay, we decided to just throw an error in both `search_modkey` and `search_fedkey `, see https://github.com/cms-sw/cmssw/pull/48137/commits/dafa02849c285fcd2e4ba132baf0a21f0d1b1c59.